thriftclient/client_pool.go

```diff
@@ -0,0 +1,211 @@
+	ReleaseClient(Client)
+}
+
+// AddressGenerator defines a function that returns the address of a thrift
+// service.
+type AddressGenerator func() string
+
+// ClientFactory defines a function that builds a Client object using a
+// the thrift primitives required to create a thrift.TClient.
+type ClientFactory func(thrift.TTransport, thrift.TProtocolFactory) Client
+
+// SingleAddressGenerator returns an AddressGenerator that always returns addr.
+func SingleAddressGenerator(addr string) AddressGenerator {
+	return func() string {
+		return addr
+	}
+}
+
+// MonitoredTTLClientFactory returns a ClientFactory that creates TTLClients
+// with a MonitoredClient as the underlying thrift.TClient.
+func MonitoredTTLClientFactory(ttl time.Duration) ClientFactory {
+	return func(trans thrift.TTransport, protoFactory thrift.TProtocolFactory) Client {
+		c := NewMonitoredClientFromFactory(trans, protoFactory)
+		return NewTTLClient(trans, c, ttl)
+	}
+}
+
+// UnmonitoredTTLClientFactory returns a ClientFactory that creates TTLClients
+// with a thrift.TStandardClient as the underlying thrift.TClient.
+func UnmonitoredTTLClientFactory(ttl time.Duration) ClientFactory {
+	return func(trans thrift.TTransport, protoFactory thrift.TProtocolFactory) Client {
+		c := thrift.NewTStandardClient(
+			protoFactory.GetProtocol(trans),
+			protoFactory.GetProtocol(trans),
+		)
+		return NewTTLClient(trans, c, ttl)
+	}
+}
+
+func newClient(socketTimeout time.Duration, genAddr AddressGenerator, clientFact ClientFactory, protoFact thrift.TProtocolFactory) (Client, error) {
+	trans, err := thrift.NewTSocketTimeout(genAddr(), socketTimeout)
+	if err != nil {
+		return nil, err
+	}
+	err = trans.Open()
+	if err != nil {
+		return nil, err
+	}
+	return clientFact(trans, protoFact), nil
+}
+
+func reportPoolStats(ctx context.Context, prefix string, pool clientpool.Pool, tickerDuration time.Duration) {
+	gauge := metricsbp.M.Gauge(prefix + ".pool-active-connections")
+	ticker := time.NewTicker(tickerDuration)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			gauge.Set(float64(pool.NumActiveClients()))
+		}
+	}
+}
+
+// NewTTLClientPool creates a ClientPool that can be used to create monitored
+// TTLClients.
+func NewTTLClientPool(ttl time.Duration, cfg ClientPoolConfig) (ClientPool, error) {
+	return newClientPool(
+		cfg,
+		SingleAddressGenerator(cfg.Addr),
+		MonitoredTTLClientFactory(ttl),
+		thrift.NewTHeaderProtocolFactory(),
+	)
+}
+
+// NewCustomClientPool creates a ClientPool that uses a custom AddressGenerator
+// and ClientFactory.
+//
+// Most services will want to just use NewTTLClientPool, this has been provided
+// to support services that have non-standard and/or legacy needs.
+func NewCustomClientPool(cfg ClientPoolConfig, genAddr AddressGenerator, clientFact ClientFactory, protoFact thrift.TProtocolFactory) (ClientPool, error) {
+	if cfg.Addr != "" {
+		log.Warnf(
```

Comment:
nit: When using zap logger interface, prefer *w version over *f version for performance reasons. when using *w version you just put a static string as the first arg, then key/value pairs follow for additional info.